### ISSUE TITLE
Alacritty window on current workspace is focused instead of changing workspaces

### ISSUE THREAD

**Issue by sangaline** [nikitabobko/AeroSpace#247]
If I have two workspaces with an alacritty window open on them and the alacritty window is not focused on the current workspace, then trying to switch to the other workspace will focus alacritty on the current workspace instead of switching workspaces. After alacritty is focused, trying to switch workspaces a second time will succeed. This behavior seems to be fairly robust against different layouts or how many windows I have open on each workspace.

Note that I'm opening the terminals with:

```bash
alt-enter = 'exec-and-forget alacritty msg create-window || open -na alacritty'
```

so all of the terminal windows correspond to one instance of the app.

Side note: opening alacritty with a stack layout will not focus it or place it in the foreground; this behavior is similar to #88.

**Comment by nikitabobko** [comment]
I suspect that your workspaces are on different monitors

If so, it's a bug in Apple API. See #101 

Workaround is to disable "Displays have separate spaces"

**Comment by sangaline** [comment]
I'm not using an external monitor, I only have the single space on the laptop screen.

**Comment by nikitabobko** [comment]
Can't reproduce

1. By "workspaces" you mean AeroSpace "workspaces" not macOS native "Spaces", right
2. Is it reproducible only with Alacritty or other multi-window apps?

**Comment by sangaline** [comment]
> By "workspaces" you mean AeroSpace "workspaces" not macOS native "Spaces", right?

Yes. I have only one native macOS space, this happens when switching AeroSpace workspaces.

> Is it reproducible only with Alacritty or other multi-window apps?

Only Alacritty.


-----

> Can't reproduce

Is there anything I can do to provide additional information that would be helpful?

* Aerospace version: 0.11.0-Beta 63e4a40bbb33f7c1b6538b6da07e385252da3db9
* MacOS version: 14.5
* Alacritty Version: 0.13.2

Here is the `aerospace debug-windows` output when I have alacritty and a finder window open on workspace 6. I run the command in alacritty, switch to the finder window, and then attempt to switch to workspace 1 (which has additional alacritty windows). Instead of switching to workspace 1, it switches focus to alacritty on workspace 6.

```
aerospace debug-windows
org.alacritty.window.21127 windowId: 21127
org.alacritty.window.21127 workspace: 6
org.alacritty.window.21127 treeNodeParent: AppBundle.TilingContainer
org.alacritty.window.21127 recognizedAsDialog: false
org.alacritty.window.21127 AXTitle: Optional(Alacritty)
org.alacritty.window.21127 AXRole: Optional(AXWindow)
org.alacritty.window.21127 AXSubrole: Optional(AXStandardWindow)
org.alacritty.window.21127 AXFocused: Optional(1)
org.alacritty.window.21127 AXFullScreen: Optional(0)
org.alacritty.window.21127 AXFrame: Optional(<AXValue 0x600002758e00> {value = x:0.000000 y:68.000000 w:1512.000000 h:913.000000 type = kAXValueCGRectType})
org.alacritty.window.21127 AXPosition: Optional(<AXValue 0x600003c4e760> {value = x:0.000000 y:68.000000 type = kAXValueCGPointType})
org.alacritty.window.21127 AXGrowArea: nil
org.alacritty.window.21127 AXMinimizeButton: nil
org.alacritty.window.21127 AXDocument: nil
org.alacritty.window.21127 AXSections: [
org.alacritty.window.21127 ]
org.alacritty.window.21127 AXCloseButton: nil
org.alacritty.window.21127 AXMain: Optional(1)
org.alacritty.window.21127 AXActivationPoint: Optional(<AXValue 0x600003cca460> {value = x:-1.000000 y:983.000000 type = kAXValueCGPointType})
org.alacritty.window.21127 AXFullScreenButton: nil
org.alacritty.window.21127 AXProxy: nil
org.alacritty.window.21127 AXDefaultButton: nil
org.alacritty.window.21127 AXMinimized: Optional(0)
org.alacritty.window.21127 AXParent: Optional(<AXUIElement Application 0x600003c57060> {pid=10046})
org.alacritty.window.21127 AXTitleUIElement: nil
org.alacritty.window.21127 AXCancelButton: nil
org.alacritty.window.21127 AXModal: Optional(0)
org.alacritty.window.21127 AXZoomButton: nil
org.alacritty.window.21127 AXSize: Optional(<AXValue 0x600003c4f1b0> {value = w:1512.000000 h:913.000000 type = kAXValueCGSizeType})
org.alacritty.window.21127 AXToolbarButton: nil
org.alacritty.window.21127 Ignored: AXChildrenInNavigationOrder, AXChildren, AXRoleDescription
org.alacritty              AXRole: Optional(AXApplication)
org.alacritty              AXTitle: Optional(alacritty)
org.alacritty              AXFunctionRowTopLevelElements: [
org.alacritty              ]
org.alacritty              AXFrame: nil
org.alacritty              AXFocusedUIElement: AXUIElement(windowId=21127, title="Alacritty", role="AXWindow", subrole="AXStandardWindow")
org.alacritty              AXFrontmost: Optional(1)
org.alacritty              AXExtrasMenuBar: nil
org.alacritty              AXMainWindow: AXUIElement(windowId=21127, title="Alacritty", role="AXWindow", subrole="AXStandardWindow")
org.alacritty              AXFocusedWindow: AXUIElement(windowId=21127, title="Alacritty", role="AXWindow", subrole="AXStandardWindow")
org.alacritty              AXMenuBar: Optional(<AXUIElement 0x600003c57060> {pid=10046})
org.alacritty              AXWindows: [
org.alacritty                  AXUIElement(windowId=21127, title="Alacritty", role="AXWindow", subrole="AXStandardWindow"),
org.alacritty                  AXUIElement(windowId=3001, title="Alacritty", role="AXWindow", subrole="AXStandardWindow"),
org.alacritty                  AXUIElement(windowId=21066, title="Alacritty", role="AXWindow", subrole="AXStandardWindow"),
org.alacritty                  AXUIElement(windowId=2994, title="Alacritty", role="AXWindow", subrole="AXStandardWindow"),
org.alacritty                  AXUIElement(windowId=1955, title="tenantbase@1f7793fbbaa8: /workspace", role="AXWindow", subrole="AXStandardWindow"),
org.alacritty                  AXUIElement(windowId=2125, title="root@30358e315dda: /forge", role="AXWindow", subrole="AXStandardWindow"),
org.alacritty                  AXUIElement(windowId=4101, title="root@ce79d7f95590: /forge", role="AXWindow", subrole="AXStandardWindow")
org.alacritty              ]
org.alacritty              AXSize: nil
org.alacritty              AXPosition: nil
org.alacritty              Ignored: AXChildren, AXChildrenInNavigationOrder, AXEnhancedUserInterface, AXPreferredLanguage, AXRoleDescription, AXHidden

com.apple.finder.window.21125 windowId: 21125
com.apple.finder.window.21125 workspace: 6
com.apple.finder.window.21125 treeNodeParent: AppBundle.TilingContainer
com.apple.finder.window.21125 recognizedAsDialog: false
com.apple.finder.window.21125 AXTitle: Optional(Downloads)
com.apple.finder.window.21125 AXRole: Optional(AXWindow)
com.apple.finder.window.21125 AXSubrole: Optional(AXStandardWindow)
com.apple.finder.window.21125 AXIdentifier: Optional(FinderWindow)
com.apple.finder.window.21125 AXFocused: Optional(0)
com.apple.finder.window.21125 AXFullScreen: Optional(0)
com.apple.finder.window.21125 AXFrame: Optional(<AXValue 0x600002746800> {value = x:0.000000 y:38.000000 w:1512.000000 h:913.000000 type = kAXValueCGRectType})
com.apple.finder.window.21125 AXPosition: Optional(<AXValue 0x600003cf4090> {value = x:0.000000 y:38.000000 type = kAXValueCGPointType})
com.apple.finder.window.21125 AXGrowArea: nil
com.apple.finder.window.21125 AXMinimizeButton: AXUIElement {
com.apple.finder.window.21125      AXRole: Optional(AXButton)
com.apple.finder.window.21125      AXTitle: nil
com.apple.finder.window.21125      AXSubrole: Optional(AXMinimizeButton)
com.apple.finder.window.21125      AXEnabled: Optional(1)
com.apple.finder.window.21125      AXParent: AXUIElement(windowId=21125, title="Downloads", role="AXWindow", subrole="AXStandardWindow")
com.apple.finder.window.21125      AXTopLevelUIElement: AXUIElement(windowId=21125, title="Downloads", role="AXWindow", subrole="AXStandardWindow")
com.apple.finder.window.21125      AXWindow: AXUIElement(windowId=21125, title="Downloads", role="AXWindow", subrole="AXStandardWindow")
com.apple.finder.window.21125      Ignored: AXFrame, AXSize, AXFocused, AXHelp, AXPosition, AXRoleDescription
com.apple.finder.window.21125 }
com.apple.finder.window.21125 AXDocument: nil
com.apple.finder.window.21125 AXSections: [
com.apple.finder.window.21125         Optional({
com.apple.finder.window.21125             SectionDescription = Toolbar;
com.apple.finder.window.21125             SectionObject = "<AXUIElement 0x600003cc9860> {pid=686}";
com.apple.finder.window.21125             SectionUniqueID = AXToolbar;
com.apple.finder.window.21125         }),
com.apple.finder.window.21125         Optional({
com.apple.finder.window.21125             SectionDescription = "Content Navigator";
com.apple.finder.window.21125             SectionObject = "<AXUIElement 0x600003cca4f0> {pid=686}";
com.apple.finder.window.21125             SectionUniqueID = AXContentNavigator;
com.apple.finder.window.21125         }),
com.apple.finder.window.21125         Optional({
com.apple.finder.window.21125             SectionDescription = Search;
com.apple.finder.window.21125             SectionObject = "<AXUIElement 0x600003cc9a40> {pid=686}";
com.apple.finder.window.21125             SectionUniqueID = AXSearch;
com.apple.finder.window.21125         }),
com.apple.finder.window.21125         Optional({
com.apple.finder.window.21125             SectionObject = "<AXUIElement 0x600003ccb780> {pid=686}";
com.apple.finder.window.21125             SectionUniqueID = AXContent;
com.apple.finder.window.21125         }),
com.apple.finder.window.21125         Optional({
com.apple.finder.window.21125             SectionObject = "<AXUIElement 0x600003ccbea0> {pid=686}";
com.apple.finder.window.21125             SectionUniqueID = AXTopLevelNavigator;
com.apple.finder.window.21125         })
com.apple.finder.window.21125 ]
com.apple.finder.window.21125 AXCloseButton: AXUIElement {
com.apple.finder.window.21125      AXRole: Optional(AXButton)
com.apple.finder.window.21125      AXTitle: nil
com.apple.finder.window.21125      AXSubrole: Optional(AXCloseButton)
com.apple.finder.window.21125      AXEnabled: Optional(1)
com.apple.finder.window.21125      AXParent: AXUIElement(windowId=21125, title="Downloads", role="AXWindow", subrole="AXStandardWindow")
com.apple.finder.window.21125      AXTopLevelUIElement: AXUIElement(windowId=21125, title="Downloads", role="AXWindow", subrole="AXStandardWindow")
com.apple.finder.window.21125      AXWindow: AXUIElement(windowId=21125, title="Downloads", role="AXWindow", subrole="AXStandardWindow")
com.apple.finder.window.21125      Ignored: AXFrame, AXSize, AXFocused, AXHelp, AXPosition, AXRoleDescription, AXEdited
com.apple.finder.window.21125 }
com.apple.finder.window.21125 AXMain: Optional(1)
com.apple.finder.window.21125 AXActivationPoint: Optional(<AXValue 0x600003c4e400> {value = x:10.000000 y:52.000000 type = kAXValueCGPointType})
com.apple.finder.window.21125 AXFullScreenButton: AXUIElement {
com.apple.finder.window.21125      AXRole: Optional(AXButton)
com.apple.finder.window.21125      AXTitle: nil
com.apple.finder.window.21125      AXSubrole: Optional(AXFullScreenButton)
com.apple.finder.window.21125      AXEnabled: Optional(1)
com.apple.finder.window.21125      AXParent: AXUIElement(windowId=21125, title="Downloads", role="AXWindow", subrole="AXStandardWindow")
com.apple.finder.window.21125      AXTopLevelUIElement: AXUIElement(windowId=21125, title="Downloads", role="AXWindow", subrole="AXStandardWindow")
com.apple.finder.window.21125      AXWindow: AXUIElement(windowId=21125, title="Downloads", role="AXWindow", subrole="AXStandardWindow")
com.apple.finder.window.21125      Ignored: AXFrame, AXSize, AXFocused, AXChildren, AXHelp, AXPosition, AXRoleDescription
com.apple.finder.window.21125 }
com.apple.finder.window.21125 AXProxy: AXUIElement(windowId=21125, title="", role="AXImage", subrole=nil)
com.apple.finder.window.21125 AXDefaultButton: nil
com.apple.finder.window.21125 AXMinimized: Optional(0)
com.apple.finder.window.21125 AXParent: Optional(<AXUIElement Application 0x600003c4e400> {pid=686})
com.apple.finder.window.21125 AXTitleUIElement: AXUIElement(windowId=21125, title=nil, role="AXStaticText", subrole=nil)
com.apple.finder.window.21125 AXCancelButton: nil
com.apple.finder.window.21125 AXModal: Optional(0)
com.apple.finder.window.21125 AXValue: nil
com.apple.finder.window.21125 AXZoomButton: AXUIElement {
com.apple.finder.window.21125      AXRole: Optional(AXButton)
com.apple.finder.window.21125      AXTitle: nil
com.apple.finder.window.21125      AXSubrole: Optional(AXFullScreenButton)
com.apple.finder.window.21125      AXEnabled: Optional(1)
com.apple.finder.window.21125      AXParent: AXUIElement(windowId=21125, title="Downloads", role="AXWindow", subrole="AXStandardWindow")
com.apple.finder.window.21125      AXTopLevelUIElement: AXUIElement(windowId=21125, title="Downloads", role="AXWindow", subrole="AXStandardWindow")
com.apple.finder.window.21125      AXWindow: AXUIElement(windowId=21125, title="Downloads", role="AXWindow", subrole="AXStandardWindow")
com.apple.finder.window.21125      Ignored: AXFrame, AXSize, AXFocused, AXChildren, AXHelp, AXPosition, AXRoleDescription
com.apple.finder.window.21125 }
com.apple.finder.window.21125 AXSize: Optional(<AXValue 0x600003ccb9f0> {value = w:1512.000000 h:913.000000 type = kAXValueCGSizeType})
com.apple.finder.window.21125 AXToolbarButton: nil
com.apple.finder.window.21125 Ignored: AXChildrenInNavigationOrder, AXChildren, AXRoleDescription
com.apple.finder              AXRole: Optional(AXApplication)
com.apple.finder              AXTitle: Optional(Finder)
com.apple.finder              AXFunctionRowTopLevelElements: [
com.apple.finder              ]
com.apple.finder              AXValue: nil
com.apple.finder              AXFrame: Optional(<AXValue 0x600002755b00> {value = x:0.000000 y:982.000000 w:0.000000 h:0.000000 type = kAXValueCGRectType})
com.apple.finder              AXFocusedUIElement: AXUIElement(windowId=21125, title=nil, role="AXOutline", subrole=nil)
com.apple.finder              AXFrontmost: Optional(1)
com.apple.finder              AXExtrasMenuBar: nil
com.apple.finder              AXMainWindow: AXUIElement(windowId=21125, title="Downloads", role="AXWindow", subrole="AXStandardWindow")
com.apple.finder              AXFocusedWindow: AXUIElement(windowId=21125, title="Downloads", role="AXWindow", subrole="AXStandardWindow")
com.apple.finder              AXMenuBar: Optional(<AXUIElement 0x600003c4e400> {pid=686})
com.apple.finder              AXWindows: [
com.apple.finder                  AXUIElement(windowId=21125, title="Downloads", role="AXWindow", subrole="AXStandardWindow"),
com.apple.finder                  AXUIElement(windowId=675, title="Downloads", role="AXWindow", subrole="AXStandardWindow"),
com.apple.finder                  Optional(<AXUIElement 0x600003cb39c0> {pid=686})
com.apple.finder              ]
com.apple.finder              AXSize: Optional(<AXValue 0x600003cb2a30> {value = w:0.000000 h:0.000000 type = kAXValueCGSizeType})
com.apple.finder              AXPosition: Optional(<AXValue 0x600003ccb9f0> {value = x:0.000000 y:982.000000 type = kAXValueCGPointType})
com.apple.finder              Ignored: AXChildren, AXChildrenInNavigationOrder, AXEnhancedUserInterface, AXPreferredLanguage, AXRoleDescription, AXHidden

!!! DISCLAIMER !!!
!!! 'debug-windows' command is not stable API. Please don't rely on the command existence and output format !!!
!!! The only intended use case is to report bugs about incorrect windows handling !!!

Debug session finished
```

**Comment by nikitabobko** [comment]
Mine versions are:
- alacritty 0.13.2 (bb8ea18)
- macOS 14.3.1 (23D60)
- AeroSpace 0.11.0-Beta

I will try to update macOS to see if I can reproduce the problem

---

You can try to debug the AeroSpace with the following patch:
```diff
diff --git a/Sources/AppBundle/tree/MacWindow.swift b/Sources/AppBundle/tree/MacWindow.swift
index 7c66cf6..d4720b3 100644
--- a/Sources/AppBundle/tree/MacWindow.swift
+++ b/Sources/AppBundle/tree/MacWindow.swift
@@ -96,8 +96,9 @@ final class MacWindow: Window, CustomStringConvertible {
 
     @discardableResult
     override func nativeFocus() -> Bool { // todo make focus reliable: make async + active waiting
+        print("--- \(title) on workspace \(workspace?.name)")
         // Raise firstly to make sure that by that time we activate the app, particular window would be already on top
-        axWindow.raise() && macApp.nsApp.activate(options: .activateIgnoringOtherApps)
+        return axWindow.raise() && macApp.nsApp.activate(options: .activateIgnoringOtherApps)
     }
 
     override func close() -> Bool {
```

to see what window of Alacritty gets focused. `nativeFocus` function has only a single usage: `Sources/AppBundle/refresh.swift:30`

**Comment by nikitabobko** [comment]
Is the problem reproducible with default Alacritty config?

**Comment by sangaline** [comment]
Yes, I still get the same behavior after `rm ~/.config/alacritty/alacritty.toml`, `touch ~/.config/alacritty/alacritty.toml`, and restarting the application. I haven't gotten to compiling AeroSpace with that change yet, but I'll update once I have.

**Comment by nikitabobko** [comment]
I've updated macOS to 14.5 (23F79), the issue is still not reproducible

When I ask AeroSpace to focus different workspace, it focuses different workspace

It's strange that the problem is reproducible only with Alacritty

**Comment by nikitabobko** [comment]
0.11.1 comes with the fix for #101. The fix might potentially fix your Alacritty problem

**Comment by sangaline** [comment]
I unfortunately still see the same behavior with 0.11.1. I've recompiled with the diff you gave and here are the debug logs. The setup here is that there is one alacritty window on workspace 8, and then one finder window and one alacritty window on workspace 9. I start out on workspace 8, move to workspace 9, move focus from alacritty to finder in workspace 9, attempt to move to workspace 8 but instead alacritty is focused on workspace 9, then attempt to move to workspace 8 a second time and it succeeds now that alacritty was focused on workspace 9.

```
--- Alacritty on workspace Optional("9")
hideViaEmulation: Hide Window(title: 'Optional("Alacritty")', role: 'Optional("AXWindow")', subrole: 'Optional("AXStandardWindow")', identifier: 'nil', modal: 'Optional("false")', windowId: 'Optional("24111")')
--- Downloads on workspace Optional("9")
--- Alacritty on workspace Optional("8")
hideViaEmulation: Hide Window(title: 'Optional("Downloads")', role: 'Optional("AXWindow")', subrole: 'Optional("AXStandardWindow")', identifier: 'Optional("FinderWindow")', modal: 'Optional("false")', windowId: 'Optional("21125")')
hideViaEmulation: Hide Window(title: 'Optional("Alacritty")', role: 'Optional("AXWindow")', subrole: 'Optional("AXStandardWindow")', identifier: 'nil', modal: 'Optional("false")', windowId: 'Optional("32514")')
hideViaEmulation: Hide Window(title: 'Optional("Alacritty")', role: 'Optional("AXWindow")', subrole: 'Optional("AXStandardWindow")', identifier: 'nil', modal: 'Optional("false")', windowId: 'Optional("24111")')
--- Alacritty on workspace Optional("8")
hideViaEmulation: Hide Window(title: 'Optional("Downloads")', role: 'Optional("AXWindow")', subrole: 'Optional("AXStandardWindow")', identifier: 'Optional("FinderWindow")', modal: 'Optional("false")', windowId: 'Optional("21125")')
hideViaEmulation: Hide Window(title: 'Optional("Alacritty")', role: 'Optional("AXWindow")', subrole: 'Optional("AXStandardWindow")', identifier: 'nil', modal: 'Optional("false")', windowId: 'Optional("32514")')
```

It appears that window ID 24111 is the workspace 8 alacritty window and 32514 is the workspace 9 one. It's interesting that both show up with `hideViaEmulation` the first time I attempt to switch to workspace 8 and then only 32514 shows up when the switch succeeds on the second try. I don't know what else is going on behind the scenes, but I wouldn't expect 24111 to need to be hidden when switching from workspace 9 to 8.

**Comment by nikitabobko** [comment]
Thanks for you patience and providing all the info and logs! I know it's tedious.

I've managed to reproduce the issue. Turns out that "Displays have separate spaces" needs to be enabled to reproduce the problem (it was disabled on my machine). It ones again proves it to me that macOS Spaces are cursed, and they must have a problematic implementation under the hood

But additionally, I noticed that the problem is reproducible only if `window.decorations = "None"` is used in Alacritty config.

Please try to use `window.decorations = "Buttonless"` (you need to restart Alacritty for the setting to take effect).

**Just a side note:** WezTerm also has "borderless" mode (`config.window_decorations = 'NONE'` https://wezfurlong.org/wezterm/config/lua/config/window_decorations.html?h=window_decorations). If you try to use "borderless" WezTerm, you will notice that AeroSpace won't be able to resize the window, but I couldn't reproduce the "Alacritty" bug. Anyway, I think it proves to us that "borderless" may cause different quirks, and it's better not to use it

To sum up, there is one proper solution:
1. Please don't use "borderless" mode

Or two possible workarounds (if you really want to use "borderless" Alacritty):
1. Disable "Displays have separate spaces" in "System Settings"
2. Span different process of Alacritty instead of `alacritty msg create-window`

I don't think AeroSpace can do anything about it, according to the logs you provided, AeroSpace commands to focus the correct window, it's macOS & Alacritty who don't listen the command

**Comment by sangaline** [comment]
> Turns out that "Displays have separate spaces" needs to be enabled to reproduce.

Awesome, I appreciate the help and debugging! Disabling this resolves the issue with workspace switching. Terminals will still open without focus in a stacked layout, but that's less of an issue for me.

> But additionally, I noticed that the problem is reproducible only if window.decorations = "None" is used in Alacritty config.

Interesting, I was definitely observing it with an empty config that included the normal window decorations, but now that I have "displays have separate spaces" it's annoying to check because it requires restarting.
